runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobInvocation.java

```diff
@@ -200,8 +208,10 @@ public void onFailure(Throwable throwable) {}
   }
 
   /** Listen for job state changes with a {@link Consumer}. */
-  public synchronized void addStateListener(Consumer<JobState.Enum> stateStreamObserver) {
-    stateStreamObserver.accept(getState());
+  public synchronized void addStateListener(Consumer<JobStateEvent> stateStreamObserver) {
+    for (JobStateEvent event : stateHistory) {
+      stateStreamObserver.accept(event);
```

Comment:
Couldn't we remove the event once we have informed the observers?

Reply:
There may be many clients connecting to this stream, and thus many potential observers interested in receiving the full list of state history.